internal/tools/read_cypher_handler_test.go

```diff
@@ -0,0 +1,329 @@
+		}
+		if result == nil || !result.IsError {
+			t.Error("Expected error result for nil database service")
+		}
+	})
+
+	t.Run("database query execution failure", func(t *testing.T) {
+		mockDB := mocks.NewMockDatabaseService(ctrl)
+		mockDB.EXPECT().
+			GetQueryType(gomock.Any(), "INVALID CYPHER", gomock.Nil(), "testdb").
+			Return(neo4j.StatementTypeReadOnly, nil)
+		mockDB.EXPECT().
+			ExecuteReadQuery(gomock.Any(), "INVALID CYPHER", gomock.Nil(), "testdb").
+			Return(nil, errors.New("syntax error"))
+
+		deps := &ToolDependencies{
+			Config:    &config.Config{Database: "testdb"},
+			DBService: mockDB,
+		}
+
+		handler := ReadCypherHandler(deps)
+		request := mcp.CallToolRequest{
+			Params: mcp.CallToolParams{
+				Arguments: map[string]any{
+					"query": "INVALID CYPHER",
+				},
+			},
+		}
+
+		result, err := handler(context.Background(), request)
+
+		if err != nil {
+			t.Errorf("Expected no error from handler, got: %v", err)
+		}
+		if result == nil || !result.IsError {
+			t.Error("Expected error result for query execution failure")
+		}
+	})
+
+	t.Run("JSON formatting failure", func(t *testing.T) {
+		mockDB := mocks.NewMockDatabaseService(ctrl)
+		mockDB.EXPECT().
+			GetQueryType(gomock.Any(), "MATCH (n) RETURN n", gomock.Nil(), "testdb").
+			Return(neo4j.StatementTypeReadOnly, nil)
+		mockDB.EXPECT().
+			ExecuteReadQuery(gomock.Any(), "MATCH (n) RETURN n", gomock.Nil(), "testdb").
+			Return([]*neo4j.Record{}, nil)
+		mockDB.EXPECT().
+			Neo4jRecordsToJSON(gomock.Any()).
+			Return("", errors.New("JSON marshaling failed"))
+
+		deps := &ToolDependencies{
+			Config:    &config.Config{Database: "testdb"},
+			DBService: mockDB,
+		}
+
+		handler := ReadCypherHandler(deps)
+		request := mcp.CallToolRequest{
+			Params: mcp.CallToolParams{
+				Arguments: map[string]any{
+					"query": "MATCH (n) RETURN n",
+				},
+			},
+		}
+
+		result, err := handler(context.Background(), request)
+
+		if err != nil {
+			t.Errorf("Expected no error from handler, got: %v", err)
+		}
+		if result == nil || !result.IsError {
+			t.Error("Expected error result for JSON formatting failure")
+		}
+	})
+
+	t.Run("non-read query type returns error", func(t *testing.T) {
+		mockDB := mocks.NewMockDatabaseService(ctrl)
+		mockDB.EXPECT().
+			GetQueryType(gomock.Any(), "CREATE (n:Test)", gomock.Nil(), "testdb").
+			Return(neo4j.StatementTypeWriteOnly, nil)
+
+		deps := &ToolDependencies{
+			Config:    &config.Config{Database: "testdb"},
+			DBService: mockDB,
+		}
+
+		handler := ReadCypherHandler(deps)
+		request := mcp.CallToolRequest{
+			Params: mcp.CallToolParams{
+				Arguments: map[string]any{
+					"query": "CREATE (n:Test)",
+				},
+			},
+		}
+
+		result, err := handler(context.Background(), request)
+		if err != nil {
+			t.Errorf("Expected no error from handler, got: %v", err)
+		}
+		if result == nil || !result.IsError {
+			t.Error("Expected error result for non-read query type")
+		}
```

Comment:
We might want to include tests for the logged messages (log.Printf) since we utilise that to report errors to the server administrators.

Reply:
This is my opinion, but I think the value that we add by testing the log message is less than the tradeoff paid in testing, especially as it will be a part in constant movement or at least is how I envision that.